### PR TITLE
 contrib/database/sql: allow setting service name per connection

### PR DESCRIPTION
**Pull request by cgilmour**
This PR provides additional options when opening a database connection, so that the service name can be overridden.
Related issue: #325 

This replaces the "driver wrapper" with a "connector wrapper" when calling `sqltrace.Open`.
The `Open` function is expanded to take additional `OpenOption` parameters to set the service name, analytics, and analytics rate. When set, these override the values used when registering the driver with sqltrace.

The `OpenDB` function is also added to match `database/sql.OpenDB`, and also supports additional `OpenOption` parameters. If the caller uses this directly, they should pass in the DSN (data source name) with the `sqltrace.OpenWIthDataSourceName` parameter. This gets used to extract metadata which is added to the spans.

One part that is likely to trigger the "don't use reflect" reflex is a mapping maintained between driver names and the driver's type as `reflect.Type`. This mapping gets used in `OpenDB` to look up the config when the driver was registered. This was unavoidable because different instances of the driver are registered by the DB driver's init function and the caller to sqltrace.Register. Since we can't compare instances, the best alternative was to compare types. 

#### TODO

- [ ] Update examples

### CODE REVIEW

File: contrib/database/sql/option.go [hunk 1]
```diff
@@ -47,3 +47,55 @@ func WithAnalyticsRate(rate float64) RegisterOption {
...
 	}
 }
+
+type openConfig struct {
+	serviceName   string
+	analyticsRate float64
+	dsn           string
+}
+
+// OpenOption represents an option that can be passed to Open or OpenDB.
```
Comment (gbbr) [inline contrib/database/sql/option.go]: I'm fine with having multiple options structures, as it gives for better documentation, but not at the cost of prefixing the option names with `Open*`. I think that's terrible (we've already tried this in the RoundTripper and it's not one of our best). Since all of these options (except `WithDSN`) apply to both `Register` and `Open`, let's just have one global `Options` structure for the entire package. Then, `sqltrace.Option` makes a nice example of all the options of the SQL package.

It'll stay consistent with what we do in [the `net/http`](https://github.com/DataDog/dd-trace-go/blob/dda0810c69ace75c63b8f1bd64f2c300dc34effd/contrib/net/http/option.go#L20-L21) and [`google.golang.org/grpc`](https://github.com/DataDog/dd-trace-go/blob/dda0810c69ace75c63b8f1bd64f2c300dc34effd/contrib/google.golang.org/grpc/option.go#L9-L11) integrations. Plus, it's weird to technically duplicate all this code.

File: contrib/database/sql/option.go [hunk 2]
```diff
@@ -47,3 +47,55 @@ func WithAnalyticsRate(rate float64) RegisterOption {
...
+		} else {
+			cfg.analyticsRate = math.NaN()
+		}
+	}
+}
+
+// OpenWithDataSourceName allows the data source name to be provided when
+// using OpenDB and a driver.Connector.
+// The value is used to automatically set tags on spans.
+func OpenWithDataSourceName(name string) OpenOption {
```
Comment (gbbr) [inline contrib/database/sql/option.go]: Why not `WithDSN`? Do we need the long version?

File: contrib/database/sql/sql.go [hunk 3]
```diff
@@ -37,24 +48,84 @@ func Register(driverName string, driver driver.Driver, opts ...RegisterOption) {
 	if cfg.serviceName == "" {
 		cfg.serviceName = driverName + ".db"
 	}
-	sql.Register(name, &tracedDriver{
-		Driver:     driver,
-		driverName: driverName,
-		config:     cfg,
-	})
+	nameToRegisterConfig[driverName] = cfg
+	// sql.Register(driverName, driver)
```
Comment (gbbr) [inline contrib/database/sql/sql.go]: Can we remove it?

File: contrib/database/sql/sql_test.go [hunk 4]
```diff
@@ -77,3 +77,85 @@ func TestPostgres(t *testing.T) {
...
+				ext.TargetHost:      nil,
+				ext.TargetPort:      nil,
+				ext.DBUser:          nil,
+				ext.DBName:          nil,
+				ext.EventSampleRate: 0.2,
+			},
+		}
+		sqltest.RunAll(t, testConfig)
+	})
+	t.Run("WithDSN", func(t *testing.T) {
```
Comment (gbbr) [inline contrib/database/sql/sql_test.go]: Can we please have an empty line above this run and the others?

File: contrib/database/sql/sql.go [hunk 5]
```diff
@@ -12,23 +12,34 @@
...
+	"context"
 	"database/sql"
 	"database/sql/driver"
 	"errors"
+	"reflect"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql/internal"
 )
 
+var driverTypeToName = make(map[reflect.Type]string)
```
Comment (gbbr) [inline contrib/database/sql/sql.go]: This is a bit confusing and takes a while to wrap your head around. Can we simplify it somehow? Maybe a structure would be clearer?
```go
// registeredDrivers holds a registry of all drivers registered via the sqltrace package.
var registeredDrivers struct {
    // keys maps driver types to their registered names.
    keys map[reflect.Type]string
    // drivers maps keys to their registered driver.
    drivers map[string]driver.Driver
    // configs maps keys to their registered configuration.
    configs map[string]*registerConfig
}
```
Alternatively, we could include `driver.Driver` into the `registerConfig` and remove the `drivers` field from above completely.

You could also add methods to this structure to help you with registration and lookup. It would contain the logic a bit better and make the rest of the code easier to read.

File: contrib/database/sql/conn.go [hunk 6]
```diff
@@ -114,3 +120,47 @@ func (tc *tracedConn) QueryContext(ctx context.Context, query string, args []dri
...
+		// nothing to trace here.
+		// See: https://github.com/DataDog/dd-trace-go/issues/270
+		return
+	}
+	name := fmt.Sprintf("%s.query", tp.driverName)
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeSQL),
+		tracer.StartTime(startTime),
+	}
+	if tp.oc.serviceName != "" {
```
Comment (gbbr) [inline contrib/database/sql/conn.go]: This could be avoided if we just merged the `Open` config into the `Register` config upon opening.

File: contrib/database/sql/option.go (hunk 2 above)
Comment (cgilmour) [reply contrib/database/sql/option.go]: Updated this.

File: contrib/database/sql/option.go (hunk 1 above)
Comment (cgilmour) [reply contrib/database/sql/option.go]: Minor misunderstanding. Updated things to `Option`.

File: contrib/database/sql/option.go [hunk 7]
```diff
@@ -4,30 +4,37 @@ import (
...
-// RegisterOption represents an option that can be passed to Register.
-type RegisterOption func(*registerConfig)
+// Option represents an option that can be passed to Register, Open or OpenDB.
+type Option func(*config)
 
-func defaults(cfg *registerConfig) {
+type registerConfig = config
+
+// RegisterOption has been deprecated in favor of Option.
+type RegisterOption = Option
```
Comment (gbbr) [inline contrib/database/sql/option.go]: I think you did it the other way around. The new type is `Option`, and the old one is being deprecated, so please just:
```go
type Option = RegisterOption
```

File: contrib/database/sql/sql.go (hunk 5 above)
Comment (cgilmour) [reply contrib/database/sql/sql.go]: Tried it this way and it's much clearer IMO. A very helpful suggestion, thanks!

File: contrib/database/sql/conn.go (hunk 6 above)
Comment (cgilmour) [reply contrib/database/sql/conn.go]: This has been updated, and config is merged at an earlier point.

File: contrib/database/sql/sql.go (hunk 3 above)
Comment (cgilmour) [reply contrib/database/sql/sql.go]: Removed.

File: contrib/database/sql/sql_test.go (hunk 4 above)
Comment (cgilmour) [reply contrib/database/sql/sql_test.go]: Extra lines added.

File: contrib/database/sql/option.go (hunk 7 above)
Comment (cgilmour) [reply contrib/database/sql/option.go]: After we did our homework, it turns out the recommendation is `type OldAPI = NewPackage.API`.
Changed in other places for consistency.

File: contrib/database/sql/conn.go [hunk 8]
```diff
@@ -114,3 +120,47 @@ func (tc *tracedConn) QueryContext(ctx context.Context, query string, args []dri
...
+type traceParams struct {
+	rc         *registerConfig
+	oc         *openConfig
+	driverName string
+	resource   string
+	meta       map[string]string
+}
+
+// tryTrace will create a span using the given arguments, but will act as a no-op when err is driver.ErrSkip.
+func (tp *traceParams) tryTrace(ctx context.Context, resource string, query string, startTime time.Time, err error) {
```
[the remark this answers is not in the document]
Comment (gbbr) [reply contrib/database/sql/conn.go]: You don't agree with this? 😄

File: contrib/database/sql/option.go (hunk 7 above)
Comment (gbbr) [reply contrib/database/sql/option.go]: Good. Thanks for changing elsewhere too. However, I feel I should add that I don't think it's a hard rule and the example you gave is for when a type moves to a different package, which is not the case here.

File: contrib/database/sql/sql.go [hunk 9]
```diff
@@ -12,49 +12,169 @@
...
+	// keys maps driver types to their registered names.
+	keys map[reflect.Type]string
+	// drivers maps keys to their registered driver.
+	drivers map[string]driver.Driver
+	// configs maps keys to their registered configuration.
+	configs map[string]*config
+}
+
+// registered checks if the named driver has already been registered.
+func (d *driverRegistry) registered(name string) bool {
```
Comment (gbbr) [inline contrib/database/sql/sql.go]: Would `isRegistered` be a clearer name and a better indication of what the return value means?

File: contrib/database/sql/sql.go [hunk 10]
```diff
@@ -12,49 +12,169 @@
...
+type driverRegistry struct {
+	// keys maps driver types to their registered names.
+	keys map[reflect.Type]string
+	// drivers maps keys to their registered driver.
+	drivers map[string]driver.Driver
+	// configs maps keys to their registered configuration.
+	configs map[string]*config
+}
+
+// registered checks if the named driver has already been registered.
```
Comment (gbbr) [inline contrib/database/sql/sql.go]: A practice I've been perpetuating is to use the word "_reports_" when boolean values are returned. I've seen this used throughout the Go standard library too and I think it's a good way to indicate what a return value of `true` means. e.g.
```go
// isRegistered reports whether the driver having name is in the registry.
```
You may, of course, use better wording :)

File: contrib/database/sql/sql.go [hunk 11]
```diff
@@ -12,49 +12,169 @@
...
+func (d *driverRegistry) registered(name string) bool {
+	_, ok := d.configs[name]
+	return ok
+}
+
+// add adds the driver and config to the driver registy.
```
Comment (gbbr) [inline contrib/database/sql/sql.go]:
```suggestion
// add adds the driver and config to the driver registry.
```

Comment (gbbr) [reply contrib/database/sql/sql.go]: Should we rephrase this to clarify that the config and name belong to the driver? The way it's phrased now it could mean that config is something unrelated. Maybe something like:
```go
// add adds the driver with the given name and config to the registry.
```

File: contrib/database/sql/sql.go [hunk 12]
```diff
@@ -12,49 +12,169 @@
...
 	"database/sql/driver"
 	"errors"
+	"math"
+	"reflect"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql/internal"
 )
 
+// registeredDrivers holds a registry of all drivers registered via the sqltrace package.
+var registeredDrivers = &driverRegistry{
```
Comment (gbbr) [inline contrib/database/sql/sql.go]: I wonder if this would look better throughout as being called simply `drivers` instead of `registeredDrivers`. Maybe the prefix is too much?

File: contrib/database/sql/sql.go [hunk 13]
```diff
@@ -12,49 +12,169 @@
...
+	if d.registered(name) {
+		return
+	}
+	d.keys[reflect.TypeOf(driver)] = name
+	d.drivers[name] = driver
+	d.configs[name] = cfg
+}
+
+// name returns the name of the driver stored in the registry.
+func (d *driverRegistry) name(driver driver.Driver) (string, bool) {
```
Comment (gbbr) [inline contrib/database/sql/sql.go]: From what I see this isn't really used except to get a name and then pass it straight to `(*driverRegistry).config`, so maybe we can simplify by removing this method and changing `config` to be:
```go
func (d *driverRegistry) config(driver driver.Driver) (cfg *config, ok bool) {
    name, ok := d.keys[reflect.TypeOf(driver)]
    if !ok {
        return false
    }
    cfg, ok = d.configs[name]
    return cfg, ok
}
```
What do you think?

File: contrib/database/sql/sql.go (hunk 9 above)
Comment (cgilmour) [reply contrib/database/sql/sql.go]: Agreed. Changed to `isRegistered`

File: contrib/database/sql/sql.go (hunk 10 above)
Comment (cgilmour) [reply contrib/database/sql/sql.go]: I used similar wording, but especially the `reports whether`. I hadn't noticed it before, but yup - it's there a lot and is clearer than `checks if`.

File: contrib/database/sql/sql.go (hunk 12 above)
Comment (cgilmour) [reply contrib/database/sql/sql.go]: I'm mixed on this. There's already lots of references to `driver.Driver`, so it might be confusing for this specific case. `registry` could work, but sounds like it'd be a package instead of a variable.

File: contrib/database/sql/sql.go (hunk 13 above)
Comment (cgilmour) [reply contrib/database/sql/sql.go]: The name gets passed down to `traceParams` and eventually in the span name.
It could be `nameAndConfig(driver.Driver) (string, *config, bool)`.

I think it's simple enough for now.

File: contrib/database/sql/option.go (hunk 7 above)
Comment (gbbr) [reply contrib/database/sql/option.go]: I still feel that your example of the "recommended way" doesn't apply here, because we're not moving packages.

File: contrib/database/sql/sql.go (hunk 12 above)
Comment (gbbr) [reply contrib/database/sql/sql.go]: Sure, as you wish.

File: contrib/database/sql/sql.go (hunk 13 above)
Comment (gbbr) [reply contrib/database/sql/sql.go]: Oh, right. Ok, I see. I guess name could be part of config, but that'd make things a bit confusing and inconsistent. Let's leave it then. Resolving.

File: contrib/database/sql/conn.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (gbbr) [reply contrib/database/sql/conn.go]: Can we get rid of this by fixing it?

**Comment by cgilmour** [comment]
Thanks!

I put some thought into the remaining comments, before merging.
They could be adjusted/corrected in a followup PR if we change our mind.
The main feature of this PR should be made available either way.
